File: src/main/messageBroker.ts
import * as fs from 'fs'
import * as path from 'path'
import { BrowserWindow } from 'electron'
import type {
  Message,
  PairState,
  CreatePairInput,
  AgentActivity,
  ActivityPhase,
  PairResources,
  ModifiedFile
} from './types'
import { processSpawner } from './processSpawner'
import { getAgentTurnDirective, type AgentTurnResult } from './agentTurn'
import { pairGitTracker } from './pairGitTracker'
import { pairResourceMonitor } from './pairResourceMonitor'
import { resetPairStateForNewTask } from './pairTaskState'

const MAX_ITERATIONS = 9999

function generateId(): string {
  return Math.random().toString(36).substring(2, 15) + Date.now().toString(36)
}

function ensureDir(dirPath: string): void {
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true })
  }
}

function createIdleActivity(label: string = 'Idle'): AgentActivity {
  const now = Date.now()
  return {
    phase: 'idle',
    label,
    startedAt: now,
    updatedAt: now
  }
}

function updateActivity(
  activity: AgentActivity,
  phase: ActivityPhase,
  label: string,
  detail?: string
): AgentActivity {
  return {
    phase,
    label,
    detail,
    startedAt: activity.startedAt,
    updatedAt: Date.now()
  }
}

const EMPTY_RESOURCES: PairResources = {
  mentor: { cpu: 0, memMb: 0 },
  executor: { cpu: 0, memMb: 0 },
  pairTotal: { cpu: 0, memMb: 0 }
}

export class MessageBroker {
  private pairStates: Map<string, PairState> = new Map()

  async initializePair(pairId: string, input: CreatePairInput): Promise<void> {
    const pairDir = path.join(input.directory, '.pair')
    ensureDir(pairDir)

    const specData = JSON.stringify({ spec: input.spec, name: input.name }, null, 2)
    fs.writeFileSync(path.join(pairDir, 'spec.json'), specData, 'utf-8')

    const gitTracking = await pairGitTracker.detectGitRepo(input.directory)
    let modifiedFiles: ModifiedFile[] = []

    if (gitTracking.available && gitTracking.rootPath) {
      gitTracking.baseline = await pairGitTracker.captureBaseline(gitTracking.rootPath)
      modifiedFiles = await pairGitTracker.getModifiedFiles(gitTracking.rootPath, gitTracking.baseline)
    }

    const state: PairState = {
      pairId,
      directory: input.directory,
      status: 'Idle',
      iteration: 0,
      maxIterations: MAX_ITERATIONS,
      turn: 'mentor',
      mentor: { status: 'Idle', turn: 'mentor', activity: createIdleActivity('Mentor idle') },
      executor: { status: 'Idle', turn: 'executor', activity: createIdleActivity('Executor idle') },
      messages: [],
      mentorActivity: createIdleActivity('Mentor idle'),
      executorActivity: createIdleActivity('Executor idle'),
      resources: { ...EMPTY_RESOURCES },
      modifiedFiles,
      gitTracking,
      automationMode: 'full-auto',
      turnArtifacts: [],
      gitReviewAvailable: gitTracking.gitReviewAvailable ?? false
    }

    this.writeStateFile(pairId, state)
    this.pairStates.set(pairId, state)

    pairResourceMonitor.registerPair(pairId, () => this.pairStates.get(pairId))
  }

  private getPairDir(pairId: string): string | null {
    const state = this.pairStates.get(pairId)
    if (!state || !state.directory) return null
    return path.join(state.directory, '.pair')
  }

  private getStateFilePath(pairId: string): string {
    const pairDir = this.getPairDir(pairId)
    return pairDir ? path.join(pairDir, 'state.json') : ''
  }

  private writeStateFile(pairId: string, state: PairState): void {
    const statePath = this.getStateFilePath(pairId)
    if (statePath) {
      fs.writeFileSync(statePath, JSON.stringify(state, null, 2), 'utf-8')
    }
  }

  startWatching(pairId: string): void {
    const state = this.pairStates.get(pairId)
    if (!state) return

    setTimeout(() => {
      state.status = 'Mentoring'
      state.mentor.status = 'Executing'
      state.mentorActivity = updateActivity(
        state.mentorActivity,
        'thinking',
        'Analyzing task',
        'Preparing first instruction'
      )
      state.executorActivity = updateActivity(
        state.executorActivity,
        'waiting',
        'Executor standing by',
        undefined
      )
      this.notifyStateUpdate(pairId)

      const specPath = path.join(state.directory, '.pair', 'spec.json')
      let spec = 'No spec provided.'
      if (fs.existsSync(specPath)) {
        try {
          spec = JSON.parse(fs.readFileSync(specPath, 'utf-8')).spec
        } catch {
          // Ignore malformed spec files and fall back to the default message.
        }
      }

      const initialMessage = `Here is the user's task specification:\n\n${spec}\n\nPlease provide the first instruction for the Executor.`
      processSpawner.triggerTurn(pairId, 'mentor', initialMessage)
    }, 1000)
  }

  async assignTask(pairId: string, input: Pick<CreatePairInput, 'name' | 'spec'>): Promise<void> {
    const state = this.pairStates.get(pairId)
    if (!state) {
      throw new Error(`Unknown pair: ${pairId}`)
    }

    const pairDir = this.getPairDir(pairId)
    if (pairDir) {
      const specData = JSON.stringify({ spec: input.spec, name: input.name }, null, 2)
      fs.writeFileSync(path.join(pairDir, 'spec.json'), specData, 'utf-8')
    }

    const nextState = resetPairStateForNewTask(state)

    if (nextState.gitTracking.available && nextState.gitTracking.rootPath) {
      nextState.gitTracking.baseline = await pairGitTracker.captureBaseline(
        nextState.gitTracking.rootPath
      )
    }

    this.pairStates.set(pairId, nextState)
    this.writeStateFile(pairId, nextState)
    this.notifyStateUpdate(pairId)
  }

  async handleAgentResponse(
    pairId: string,
    role: 'mentor' | 'executor',
    result: AgentTurnResult
  ): Promise<void> {
    const state = this.pairStates.get(pairId)
    if (!state) return

    if (state.iteration >= state.maxIterations) {
      state.status = 'Error'
      state.mentorActivity = updateActivity(
        state.mentorActivity,
        'error',
        'Max iterations reached',
        undefined
      )
      state.executorActivity = updateActivity(
        state.executorActivity,
        'idle',
        'Executor stopped',
        undefined
      )
      this.writeStateFile(pairId, state)
      this.notifyStateUpdate(pairId)
      return
    }

    const directive = getAgentTurnDirective(role, result)

    const newMessage: Message = {
      id: generateId(),
      timestamp: Date.now(),
      from: role,
      to: directive.to,
      type: directive.messageType,
      content: directive.content.trim(),
      iteration: state.iteration
    }

    state.messages.push(newMessage)
    state[role].lastMessage = newMessage

    if (directive.type === 'pause') {
      state.status = 'Reviewing'
      state.turn = 'mentor'
      state.mentor.status = 'Reviewing'
      state.mentorActivity = updateActivity(
        state.mentorActivity,
        'idle',
        'Paused: No response from agent',
        'Check logs or retry turn'
      )
      state.executorActivity = updateActivity(
        state.executorActivity,
        'idle',
        'Executor idle',
        undefined
      )

      this.writeStateFile(pairId, state)
      this.notifyRenderer(pairId, newMessage)
      this.notifyStateUpdate(pairId)
      return
    }

    if (role === 'mentor') {
      if (directive.type === 'finish') {
        state.status = 'Finished'
        state.mentor.status = 'Finished'
        state.executor.status = 'Finished'
        state.mentorActivity = updateActivity(
          state.mentorActivity,
          'idle',
          'Task completed',
          undefined
        )
        state.executorActivity = updateActivity(
          state.executorActivity,
          'idle',
          'Executor stopped',
          undefined
        )

        this.writeStateFile(pairId, state)
        this.notifyRenderer(pairId, newMessage)
        this.notifyStateUpdate(pairId)
        return
      }

      state.status = 'Executing'
      state.turn = 'executor'
      state.iteration++
      state.mentor.status = 'Idle'
      state.executor.status = 'Executing'
      state.mentorActivity = updateActivity(
        state.mentorActivity,
        'waiting',
        'Waiting for executor',
        undefined
      )
      state.executorActivity = updateActivity(
        state.executorActivity,
        'thinking',
        'Executing instruction',
        undefined
      )

      this.writeStateFile(pairId, state)
      this.notifyRenderer(pairId, newMessage)
      this.notifyStateUpdate(pairId)

      processSpawner.triggerTurn(pairId, 'executor', directive.content)
    } else {
      if (
        state.gitTracking.available &&
        state.gitTracking.rootPath &&
        state.gitTracking.baseline
      ) {
        state.modifiedFiles = await pairGitTracker.getModifiedFiles(
          state.gitTracking.rootPath,
          state.gitTracking.baseline
        )
      }

      state.status = 'Reviewing'
      state.turn = 'mentor'
      state.executor.status = 'Idle'
      state.mentor.status = 'Reviewing'
      state.executorActivity = updateActivity(
        state.executorActivity,
        'waiting',
        'Executor completed',
        undefined
      )
      state.mentorActivity = updateActivity(
        state.mentorActivity,
        'thinking',
        'Reviewing executor output',
        undefined
      )

      this.writeStateFile(pairId, state)
      this.notifyRenderer(pairId, newMessage)
      this.notifyStateUpdate(pairId)

      processSpawner.triggerTurn(pairId, 'mentor', `Executor summary:\n${directive.content}`)
    }
  }

  humanFeedback(pairId: string, approved: boolean): void {
    const state = this.pairStates.get(pairId)
    if (!state) return

    const message: Message = {
      id: generateId(),
      timestamp: Date.now(),
      from: 'human',
      to: 'mentor', // Direct human feedback goes to mentor to adjust the plan
      type: 'feedback',
      content: approved
        ? 'The user approved the results. You may finalize the task.'
        : 'The user rejected the results. Please review and provide a new instruction to the Executor to fix the issues.',
      iteration: state.iteration
    }

    state.messages.push(message)
    state.iteration++

    if (approved) {
      state.status = 'Finished'
      state.mentor.status = 'Finished'
      state.executor.status = 'Finished'
      this.writeStateFile(pairId, state)
      this.notifyRenderer(pairId, message)
      this.notifyStateUpdate(pairId)
    } else {
      state.status = 'Mentoring'
      state.turn = 'mentor'
      state.mentor.status = 'Reviewing'
      state.executor.status = 'Idle'
      this.writeStateFile(pairId, state)
      this.notifyRenderer(pairId, message)
      this.notifyStateUpdate(pairId)

      processSpawner.triggerTurn(pairId, 'mentor', message.content)
    }
  }

  getMessages(pairId: string): Message[] {
    return this.pairStates.get(pairId)?.messages || []
  }

  getState(pairId: string): PairState | undefined {
    return this.pairStates.get(pairId)
  }

  async stopPair(pairId: string): Promise<void> {
    const state = this.pairStates.get(pairId)
    
    if (state?.gitTracking.available && state.gitTracking.rootPath) {
      const modifiedFiles = await pairGitTracker.getModifiedFiles(
        state.gitTracking.rootPath,
        state.gitTracking.baseline || ''
      )
      
      if (modifiedFiles.length > 0) {
        const fileList = modifiedFiles
          .map(f => `  ${f.status} ${f.displayPath}`)
          .join('\n')
        
        const message: Message = {
          id: generateId(),
          timestamp: Date.now(),
          from: 'human',
          to: 'human',
          type: 'result',
          content: `\n📝 Session ended. Please review the following changes:\n\n${fileList}\n\nThese changes are uncommitted. Use git to commit them when ready.`,
          iteration: state.iteration
        }
        
        state.messages.push(message)
        this.writeStateFile(pairId, state)
        this.notifyRenderer(pairId, message)
      }
    }
    
    pairResourceMonitor.unregisterPair(pairId)
    this.pairStates.delete(pairId)
  }

  retryTurn(pairId: string): void {
    const state = this.pairStates.get(pairId)
    if (!state) return

    if (state.status !== 'Error') return

    const lastMessage = state.messages[state.messages.length - 1]
    if (!lastMessage) return

    state.status = state.turn === 'mentor' ? 'Mentoring' : 'Executing'
    state.mentorActivity = updateActivity(
      state.mentorActivity,
      'thinking',
      'Retrying turn',
      undefined
    )
    state.executorActivity = updateActivity(
      state.executorActivity,
      'thinking',
      'Retrying turn',
      undefined
    )

    this.writeStateFile(pairId, state)
    this.notifyStateUpdate(pairId)

    processSpawner.triggerTurn(pairId, state.turn, lastMessage.content)
  }

  updateActivity(
    pairId: string,
    role: 'mentor' | 'executor',
    phase: ActivityPhase,
    label: string,
    detail?: string
  ): void {
    const state = this.pairStates.get(pairId)
    if (!state) return

    const activity = role === 'mentor' ? state.mentorActivity : state.executorActivity
    const updated = updateActivity(activity, phase, label, detail)

    if (role === 'mentor') {
      state.mentorActivity = updated
      state.mentor.activity = updated
    } else {
      state.executorActivity = updated
      state.executor.activity = updated
    }

    this.notifyStateUpdate(pairId)
  }

  updateResources(pairId: string, resources: PairResources): void {
    const state = this.pairStates.get(pairId)
    if (!state) return
    state.resources = resources
    this.notifyStateUpdate(pairId)
  }

  private notifyRenderer(pairId: string, message: Message): void {
    BrowserWindow.getAllWindows().forEach((win) => {
      win.webContents.send('pair:message', { pairId, message })
    })
  }

  private notifyStateUpdate(pairId: string): void {
    const state = this.pairStates.get(pairId)
    if (!state) return
    BrowserWindow.getAllWindows().forEach((win) => {
      win.webContents.send('pair:state', {
        pairId,
        status: state.status,
        iteration: state.iteration,
        maxIterations: state.maxIterations,
        turn: state.turn,
        mentorStatus: state.mentor.status,
        executorStatus: state.executor.status,
        mentorActivity: state.mentorActivity,
        executorActivity: state.executorActivity,
        resources: state.resources,
        modifiedFiles: state.modifiedFiles,
        gitTracking: state.gitTracking,
        automationMode: state.automationMode
      })
    })
  }
}

export const messageBroker = new MessageBroker()

File: src-tauri/src/message_broker.rs
use crate::types::{
    ActivityPhase, AgentActivity, AgentRole, Message, MessageSender, MessageType, PairResources,
    PairState, PairStatus, ResourceInfo, CreatePairInput, GitTracking, AgentState
};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use tauri::{AppHandle, Emitter};
use std::time::{SystemTime, UNIX_EPOCH};

pub struct MessageBroker {
    pair_states: Arc<Mutex<HashMap<String, PairState>>>,
    app_handle: Option<AppHandle>,
}

impl MessageBroker {
    pub fn new() -> Self {
        Self {
            pair_states: Arc::new(Mutex::new(HashMap::new())),
            app_handle: None,
        }
    }

    pub fn set_app_handle(&mut self, handle: AppHandle) {
        self.app_handle = Some(handle);
    }

    fn now() -> u64 {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64
    }

    fn create_idle_activity(label: &str) -> AgentActivity {
        let now = Self::now();
        AgentActivity {
            phase: ActivityPhase::Idle,
            label: label.to_string(),
            detail: None,
            started_at: now,
            updated_at: now,
        }
    }

    pub fn initialize_pair(&self, pair_id: &str, input: CreatePairInput) -> Result<(), String> {
        println!("[MessageBroker::initialize_pair] Initializing pair: {}", pair_id);
        
        let empty_resources = PairResources {
            mentor: ResourceInfo { cpu: 0.0, mem_mb: 0.0 },
            executor: ResourceInfo { cpu: 0.0, mem_mb: 0.0 },
            pair_total: ResourceInfo { cpu: 0.0, mem_mb: 0.0 },
        };

        let state = PairState {
            pair_id: pair_id.to_string(),
            directory: input.directory.clone(),
            status: PairStatus::Idle,
            iteration: 0,
            max_iterations: 9999,
            turn: AgentRole::Mentor,
            mentor: AgentState {
                status: PairStatus::Idle,
                turn: AgentRole::Mentor,
                last_message: None,
                activity: Self::create_idle_activity("Mentor idle"),
            },
            executor: AgentState {
                status: PairStatus::Idle,
                turn: AgentRole::Executor,
                last_message: None,
                activity: Self::create_idle_activity("Executor idle"),
            },
            messages: Vec::new(),
            mentor_activity: Self::create_idle_activity("Mentor idle"),
            executor_activity: Self::create_idle_activity("Executor idle"),
            resources: empty_resources,
            modified_files: Vec::new(),
            git_tracking: GitTracking {
                available: false,
                root_path: None,
                baseline: None,
                git_review_available: Some(false),
            },
            automation_mode: "full-auto".to_string(),
            git_review_available: false,
        };

        let mut pair_states = self.pair_states.lock().unwrap();
        pair_states.insert(pair_id.to_string(), state);
        println!("[MessageBroker::initialize_pair] Pair state inserted successfully");
        Ok(())
    }

    #[allow(dead_code)]
    pub fn add_message(&self, pair_id: &str, message: Message) {
        let mut pair_states = self.pair_states.lock().unwrap();
        if let Some(state) = pair_states.get_mut(pair_id) {
            state.messages.push(message.clone());
            
            if message.msg_type == MessageType::Handoff {
                state.turn = if matches!(state.turn, AgentRole::Mentor) {
                    AgentRole::Executor
                } else {
                    AgentRole::Mentor
                };
                state.iteration += 1;
            }

            self.notify_state_update(pair_id, state);
            
            if let Some(handle) = &self.app_handle {
                let _ = handle.emit("pair:message", message);
            }
        }
    }

    pub fn add_log_line(&self, pair_id: &str, role: &str, line: &str) {
        let mut pair_states = self.pair_states.lock().unwrap();
        if let Some(state) = pair_states.get_mut(pair_id) {
            let msg = Message {
                id: uuid::Uuid::new_v4().to_string(),
                timestamp: Self::now(),
                from: if role == "mentor" { MessageSender::Mentor } else { MessageSender::Executor },
                to: "human".to_string(),
                msg_type: MessageType::Progress,
                content: line.to_string(),
                iteration: state.iteration,
            };
            
            state.messages.push(msg.clone());
            
            if let Some(handle) = &self.app_handle {
                let _ = handle.emit("pair:message", msg);
                let _ = handle.emit("pair:state", state.clone());
            }
        }
    }

    pub fn start_watching(&self, pair_id: &str, active_processes: Arc<Mutex<HashMap<String, tokio::process::Child>>>) {
        let mut pair_states = self.pair_states.lock().unwrap();
        if let Some(state) = pair_states.get_mut(pair_id) {
            state.status = PairStatus::Mentoring;
            state.turn = AgentRole::Mentor;
            state.iteration = 1;
            
            state.mentor.status = PairStatus::Executing;
            state.mentor_activity.phase = ActivityPhase::Thinking;
            state.mentor_activity.label = "Analyzing task".to_string();
            state.mentor_activity.detail = Some("Preparing first instruction".to_string());
            state.mentor_activity.updated_at = Self::now();
            
            state.executor_activity.phase = ActivityPhase::Waiting;
            state.executor_activity.label = "Executor standing by".to_string();
            state.executor_activity.updated_at = Self::now();

            self.notify_state_update(pair_id, state);
            
            println!("[MessageBroker] Started watching pair {}", pair_id);
        }
        drop(pair_states);

        let pair_states_clone = self.pair_states.clone();
        let app_handle_clone = self.app_handle.clone();
        let pair_id_string = pair_id.to_string();
        let active_processes_clone = active_processes.clone();

        tauri::async_runtime::spawn(async move {
            let mut sys = sysinfo::System::new_all();
            loop {
                tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
                let mut guard = pair_states_clone.lock().unwrap();
                if let Some(state) = guard.get_mut(&pair_id_string) {
                    if matches!(state.status, PairStatus::Finished | PairStatus::Error | PairStatus::Idle) {
                        break;
                    }
                    
                    crate::git_tracker::GitTracker::update_state(state);
                    crate::resource_monitor::ResourceMonitor::update_state(state, &mut sys, active_processes_clone.clone());
                    
                    if let Some(handle) = &app_handle_clone {
                        let _ = handle.emit("pair:state", state.clone());
                    }
                } else {
                    break;
                }
            }
        });
    }

    pub fn get_state(&self, pair_id: &str) -> Option<PairState> {
        let pair_states = self.pair_states.lock().unwrap();
        pair_states.get(pair_id).cloned()
    }

    pub fn update_agent_activity(&self, pair_id: &str, role: &str, phase: crate::types::ActivityPhase, label: String, detail: Option<String>) {
        let mut pair_states = self.pair_states.lock().unwrap();
        if let Some(state) = pair_states.get_mut(pair_id) {
             let activity = if role == "mentor" {
                 &mut state.mentor_activity
             } else {
                 &mut state.executor_activity
             };
             
             activity.phase = phase;
             activity.label = label;
             activity.detail = detail;
             activity.updated_at = Self::now();
             
             self.notify_state_update(pair_id, state);
        }
    }

    fn notify_state_update(&self, _pair_id: &str, state: &PairState) {
        if let Some(handle) = &self.app_handle {
            let _ = handle.emit("pair:state", state);
        }
    }
}
